### PR TITLE
Quality of life and helper callback functions

### PR DESCRIPTION
**Pull request by laserkelvin**
This PR introduces and adds a bunch of changes pertaining to informing the user of things happening under the hood, particularly during training.

One of the big philosophical changes is also focusing more on enabling logging to be done with `TensorBoardLogger` and `WandbLogger`s by writing functions more tailored to them, rather than before where loggers were treated in the abstract entirely.

## Summary

- Changed the use of coordinates in periodic boundary utilities to use cartesian coordinates, not fractional coordinates. Also included a warning message that looks at the coordinates as part of diagnostics.
- For model training, task modules now include `log_embeddings` and `log_embeddings_every_n_steps` arguments that are saved to `hparams`, which as the pair suggests, allow you to regularly log embedding vectors for analysis. This will let you ensure oversmoothing doesn't occur, where all of the embedding features become identical.
- Introduced a `TrainingHelperCallback`, which is intended to help diagnose some common issues with training, such as unused parameters, missing gradients, tiny gradients, etc. Complimentary to the change above, there is an option to inject a forward hook to any encoder (assuming it produces an `Embeddings` structure), and uses it to calculate the variance in embeddings.
- Introduced a `ModelAutocorrelation` callback, which will perform an autocorrelation analysis on model parameters and gradients over the course of training. Basically this gives you some insight into how the training dynamics appear, i.e. too much correlation = probably not good.

My intention for the `TrainingHelperCallback` is to be like a guide for best practices: we can refine this as we go and discover new things, and hopefully will be useful for everyone including new users.

### CODE REVIEW

**Comment by laserkelvin** [comment]
I think I have a lead on what the issue is: because of how `SAM` works, and because of the modifications to "stashing" embeddings in the batch structure, we now end up with two disjoint computational graphs that causes `backward` to break.

This needs a bit of thought to fix...

**Comment by laserkelvin** [comment]
Confirming this by changing out the `BaseTaskModule.forward`:

```python
        if "embeddings" in batch:
            embeddings = batch.get("embeddings")
        else:
            embeddings = self.encoder(batch)
            batch["embeddings"] = embeddings
        outputs = self.process_embedding(embeddings)
        return outputs
```

Removing the branch, and just running the encoder + processing embeddings works (i.e. don't try and grab cached embeddings).

Ideally there would be a way to check if embeddings originated from the same computational graph, but that take a lot more surgery than this PR warrants. I'll think of an alternative to this.

File: matsciml/lightning/callbacks.py [hunk 1]
```diff
@@ -845,3 +850,563 @@ def _second_step(
...
+        for parameter in pl_module.encoder.parameters():
+            encoder_norm_vals.append(parameter.detach().norm().cpu().item())
+        for head in pl_module.output_heads.values():
+            for parameter in head.parameters():
+                output_norm_vals.append(parameter.detach().norm().cpu().item())
+        encoder_norm_vals = np.array(encoder_norm_vals)
+        output_norm_vals = np.array(output_norm_vals)
+        encoder_median = np.median(encoder_norm_vals)
+        output_median = np.median(output_norm_vals)
+        if encoder_median < (2.0 * output_median):
```
Comment (melo-gonzo) [inline matsciml/lightning/callbacks.py:1053]: Another param that should probably be set in the module initialization.

File: matsciml/lightning/callbacks.py [hunk 2]
```diff
@@ -845,3 +850,563 @@ def _second_step(
...
+            python_logger.warning(
+                "Median encoder weights are significantly smaller than output heads:"
+                " encoder median norm: {encoder_median:.3e},"
+                " output head: {output_median:.3e}"
+            )
+        # optionally record to service as well
+        if log_history and pl_module.logger is not None:
+            log_service = pl_module.logger.experiment
```
Comment (melo-gonzo) [inline matsciml/lightning/callbacks.py:1061]: Probably not a huge deal, but when multiple loggers are used this will default to the first logger entered, for example loggers=[csv_logger, wandb_logger], `pl_module.logger` would be csv_logger in this case. Again, probably unlikely to encounter this in the wild - just making note. (I use csv logger + wandb 🙂). [loggers code](https://github.com/Lightning-AI/pytorch-lightning/blob/06ea3a05716a6d1f4a96cfb25021accdd18d8146/src/lightning/pytorch/trainer/trainer.py#L1575)

**Review by melo-gonzo — COMMENTED** [review]
Some great features here, thanks for doing all of this! I threw in a few comments, I know you're still working on things.

File: matsciml/lightning/callbacks.py [hunk 3]
```diff
@@ -845,3 +850,563 @@ def _second_step(
...
+    logger = getLogger("matsciml.helper")
+    logger.setLevel("INFO")
+    if isinstance(output, Embeddings):
+        # check the magnitude of both node and system level embeddings
+        if output.system_embedding is not None:
+            sys_z = output.system_embedding.detach().cpu()
+            # calculate representative statistics
+            sys_z_med = sys_z.median().item()
+            sys_z_var = sys_z.var().item()
+            if sys_z_med > 10.0:
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply matsciml/lightning/callbacks.py:885]: The number is based my own heuristics

Because this is meant to be a function registered as a forward hook, I can't really play with the arguments too easily. In principle we could make the function dynamically generated, but I don't have the bandwidth right now to refactor it...

File: matsciml/lightning/callbacks.py [hunk 4]
```diff
@@ -845,3 +850,563 @@ def _second_step(
...
+                logger.warning(
+                    f"Median system/graph embedding value is greater than 10 ({sys_z_med})"
+                )
+            if sys_z_var <= 1e-5:
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply matsciml/lightning/callbacks.py:889]: See my comment above, same applies here

File: examples/callbacks/autocorrelation.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+                cutoff_dist=20.0,
+                node_keys=["pos", "atomic_numbers"],
+            ),
+            DistancesTransform(),
+        ],
+    },
+)
+
+# tensorboard logging if working purely locally, otherwise wandb
+logger = WandbLogger(
```
[the remark this answers is not in the document]
Comment (laserkelvin) [reply examples/callbacks/autocorrelation.py:58]: I guess strictly speaking no, since these are optional functions. If a user wants to use these functions, they can install one or both (I don't want to pin them right now)

File: matsciml/lightning/callbacks.py (hunk 1 above)
Comment (laserkelvin) [reply matsciml/lightning/callbacks.py:1053]: This one again is like another hardcoded heuristic, similar reason but different since this is a `staticmethod` we don't have access to the callback's attributes unless we refactor heavily

File: matsciml/lightning/callbacks.py (hunk 2 above)
Comment (laserkelvin) [reply matsciml/lightning/callbacks.py]: Thanks for pointing that out! Addressed in aeab6cdf

**Review by melo-gonzo — APPROVED** [review]
This will bring some great utilities and helpful debugging tools! Looks good to merge.
